sips/sip-011/sip-011-catastrophic-failure-recovery.md

```diff
@@ -0,0 +1,459 @@
+example, a bug that causes all nodes in the network to become unable to respond
+to network requests is a liveness failure.
+
+The second class of bugs are _safety failures_.  These are bugs in which the
+blockchain itself permits state-transitions that are undesirable and
+inconsistent with the design of the system.  For example, a bug that would let
+anyone move anyone else's STX tokens would be a safety failure.  As another
+example, a bug in which some Clarity code's execution led to the wrong result
+(but consistent across the network) would be a safety failure.
+
+This document describes the ways in which Stacks ecosystem participants can
+coordinate to recover from various instances of these failure modes, and when it
+is appropriate to use them.  These procedures include:
+
+* Announcing a new point-release of the Foundation's reference implementation of
+  the Stacks node software
+* Covertly fixing systemically-important nodes to fix security-sensitive bugs
+* Altering or extending the p2p network protocol
+* Soft forks, which constrain what kinds of transactions may be mined on the
+  canonical fork
+* Hard forks, which are backwards-incompatible upgrades
+
+## A Note on Scope
+
+The absence of a feature is not a catastrophic failure.  The blockchain behaving
+consistently with the SIPs that describe its design is not a catastrophic
+failure, even if the behavior is not desired.  This SIP does not describe how to
+remedy these situations; it is only concerned with manual recovery procedures
+and guidelines for restoring the network after a network-wide safety or liveness
+failure is discovered.
+
+# Specification
+
+The following catastrophic failure recovery procedures are presented in order
+from least-disruptive to most-disruptive.  When choosing the recovery procedure
+to execute, ecosystem participants should err on choosing the least disruptive
+procedure that will fix the problem at hand.  A recovery procedure should only
+be taken only once all less-disruptive recovery procedures have been considered
+or tried.
+
+## Procedure for Pushing out a Critical Bugfix Point-Release
+
+In the event that a safety or liveness failure can be readily addressed in
+public _without_ altering the blockchain's social contract (i.e. no changes to
+the networking protocols, mining protocols, or chainstate validation protocols),
+the following procedure will be carried out:
+
+1. The code for the fix will be written and pushed to a Github branch prefixed
+with `fix/` to the Stacks Blockchain reference implementation, hosted at
+https://github.com/blockstack/stacks-blockchain.
+2. The branch will be submitted as a pull-request to the `master` branch, and
+will be reviewed and approved by at least two blockchain engineers, representing
+both the Stacks Foundation and at least one other major Stacks ecosystem entity.
+3. If warranted, an unofficial announcement will be made to various public
```

Comment:
When would this not be warranted? Given:

> A failure qualifies as a _catastrophic_ failure if and only if there is no conceivable way for the correct nodes in the network to make progress and preserve safety without human intervention.

Two options that could happen automatically:

- setup GitHub to announce any releases for stacks-blockchain automatically
- setup Discord to accept and relay the message from announce@stacks.org

Reply:
All the more-disruptive procedures use this procedure as a step.  Some of them, like the one below it, require a public embargo on talking about the bug because the act of making the bug public knowledge would make the problem worse.  For example, if someone discovered a bug that would let an attacker steal anyone's STX, we would _not_ announce the bug until the fix was already deployed.

I like the idea of relaying all messages from announce@stacks.org to Discord.  Is there a bot that can do this?  Not too familiar with Discord (I'm old-school), but I'd love to have a dedicated #announce channel which just relayed these messages.

Reply:
Why even have a separate (3) step? IMO step (5) could very well be re-worded as "Availability of new binaries will be communicated via standard channels per established norms of the project -- currently via the announce@ email list and Discord"

In particular, is there some specific advantage an "unofficial" announcement has? If Github releases are a source of truth, availability of a new release is automatically a "formal" announcement IMO.

Reply:
(Github didn't render your comment in the appropriate place for some reason.  Meant to post this reply here).

The point is to let node operators know when the source code for the fix is available for scrutiny and testing (which is mentioned at the end of (3)).